### PR TITLE
SOLR-11028: Created V2 equivalent of REPLACENODE

### PR DESCRIPTION
**Pull request by joshgog**
https://issues.apache.org/jira/browse/SOLR-11028

# Description

Create a v2 equivalent of 'REPLACENODE command.

# Solution

Created an API that uses JAX-RS annotations.

# Tests

 Test case for sent remote-messages is included in `ReplaceNodeApiTest`

# Checklist

Please review the following and check all that apply:

- [x] I have reviewed the guidelines for [How to Contribute](https://wiki.apache.org/solr/HowToContribute) and my code conforms to the standards described there to the best of my ability.
- [x] I have created a Jira issue and added the issue ID to my pull request title.
- [x] I have given Solr maintainers [access](https://help.github.com/en/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork) to contribute to my PR branch. (optional but recommended)
- [x] I have developed this patch against the `main` branch.
- [x] I have run `./gradlew check`.
- [x] I have added tests for my changes.
- [x] I have added documentation for the [Reference Guide](https://github.com/apache/solr/tree/main/solr/solr-ref-guide)

### CODE REVIEW

File: solr/core/src/java/org/apache/solr/handler/admin/api/ReplaceNodeAPI.java
Comment (joshgog) [inline solr/core/src/java/org/apache/solr/handler/admin/api/ReplaceNodeAPI.java:102]: Not sure if I was supposed to use CollectionParams   SOURCE_NODE and TARGET_NODE

File: solr/core/src/java/org/apache/solr/handler/admin/api/ReplaceNodeAPI.java
Comment (noblepaul) [inline solr/core/src/java/org/apache/solr/handler/admin/api/ReplaceNodeAPI.java:55]: why do we have a`/nodes` prefix? it's confusing

it should be a `/cluster` prefix  as in `cluster/nodes`

Comment (joshgog) [reply solr/core/src/java/org/apache/solr/handler/admin/api/ReplaceNodeAPI.java:55]: I used the path as documented in the proposed api design document [here](https://docs.google.com/spreadsheets/d/1HAoBBFPpSiT8mJmgNZKkZAPwfCfPvlc08m5jz3fQBpA/edit#gid=1549066254)

Comment (gerlowskija) [reply solr/core/src/java/org/apache/solr/handler/admin/api/ReplaceNodeAPI.java:54]: **TL;DR** I like Noble's suggestion to switch to `/cluster/nodes/nodeName/...`.  I'd say go with that unless you have any objections Josh?

Additional Context:

Yeah, I pointed Josh at the v2 API syntax proposed in that spreadsheet.  There was some debate there about how to handle this and other "node related" APIs.  In particular we struggled to come up with a syntax that made sense for Solr's weird mix of node-local (e.g. the v1 `/admin/info/logging`) and truly distributed (`/admin/collections?action=REPLACENODE`) APIs related to nodes.  See the previous discussion [here](https://docs.google.com/spreadsheets/d/1HAoBBFPpSiT8mJmgNZKkZAPwfCfPvlc08m5jz3fQBpA/edit?disco=AAAAbSSbh5M).

**Review by gerlowskija — COMMENTED** [review]
Was hoping to get this fully reviewed this afternoon, but ran out of time.  Publishing what I got around to but I'll try to circle back to this in the morning to finish!

File: solr/core/src/java/org/apache/solr/handler/admin/api/ReplaceNodeAPI.java
Comment (joshgog) [reply solr/core/src/java/org/apache/solr/handler/admin/api/ReplaceNodeAPI.java:71]: Sure

File: solr/core/src/java/org/apache/solr/handler/admin/api/ReplaceNodeAPI.java
Comment (joshgog) [reply solr/core/src/java/org/apache/solr/handler/admin/api/ReplaceNodeAPI.java]: This is an outdated code. The current code only accepts the targetNodeName in the request body

File: solr/core/src/java/org/apache/solr/handler/admin/api/ReplaceNodeAPI.java
Comment (joshgog) [reply solr/core/src/java/org/apache/solr/handler/admin/api/ReplaceNodeAPI.java:95]: > Unless you had a particular reason for excluding this from your RequestBody POJO below and from this message-creation here, that should probably be added?

Will add that


> Also, mostly unrelated to this PR, I notice that the [REPLACENODE ref-guide docs](https://solr.apache.org/guide/8_11/cluster-node-management.html#replacenode-parameters) mention a few parameters (parallel, timeout, etc.) that I don't see used in the code (either with or without your PR). Are those docs just plain wrong? Or am I missing something?

You are right, the docs is not in sync with the code.. To add them to the RequestBody POJO will we have to write code for their functionality e.g for parallel  parameter `If this flag is set to true, all replicas are created in separate threads`

Comment (gerlowskija) [reply solr/core/src/java/org/apache/solr/handler/admin/api/ReplaceNodeAPI.java]: OK, 😌 , thanks for the double-check on that.  Let's ignore them for now.

I'll try to do some digging to see what the story is with those params (were they removed intentionally?  accidentally?).  But whatever I find there, it'll mostly just be to record in a separate JIRA.

File: solr/core/src/java/org/apache/solr/handler/admin/CollectionsHandler.java
Comment (gerlowskija) [inline solr/core/src/java/org/apache/solr/handler/admin/CollectionsHandler.java:1802]: [0] Prior to this PR, if the user didn't explicitly specify a waitForFinalState value, the created "message" wouldn't contain that key at all.

The use of `getBool(key, defaultVal)` here means that the constructed RNRB will _always_ have a waitForFinalState value that gets added to the "message" that the v2 code builds.  This isn't a problem necessarily; the docs say that this parameter defaults to "false" - so this shouldn't change any behavior.  But it is something to keep an eye on if test failures crop up later.

If you wanted to be more conservative though, you could drop the default-val here to keep closer in line with the existing behavior.

File: solr/core/src/java/org/apache/solr/handler/admin/CollectionsHandler.java
Comment (gerlowskija) [inline solr/core/src/java/org/apache/solr/handler/admin/CollectionsHandler.java:1797]: [-1] I'm fine with dropping the deprecated old parameter names in this PR, but if we drop them here, we should also drop the code that checks for them in [ReplaceNodeCmd](https://github.com/apache/solr/blob/c99af207c761ec34812ef1cc3054eb2804b7448b/solr/core/src/java/org/apache/solr/cloud/api/collections/ReplaceNodeCmd.java#L67)

File: solr/core/src/java/org/apache/solr/handler/admin/api/ReplaceNodeAPI.java
Comment (gerlowskija) [inline solr/core/src/java/org/apache/solr/handler/admin/api/ReplaceNodeAPI.java:113]: [-1] v1 REPLACENODE supports the "async" parameter, so we should probably add that to the RequestBody as well.

(`CollectionsHandler.submitCollectionApiCommand`, which you use above, already has logic built around the parameter, so all we should have to do is add it to the RequestBody here, and then make sure it gets added to the message in `createRemoteMessage` above.)

File: solr/core/src/java/org/apache/solr/handler/admin/api/ReplaceNodeAPI.java
Comment (gerlowskija) [inline solr/core/src/java/org/apache/solr/handler/admin/api/ReplaceNodeAPI.java:98]: [0] If the user doesn't provide a request-body, e.g.

```
curl -ilk -X POST "http://localhost:8983/api/cluster/nodes/localhost:7574_solr/commands/replace"
```

then this line will cause a NullPointerException.

I'm a little torn on this - I think it'd technically be "correct" if we forced users to provide an empty JSON object (`{}`) as the request body even if they don't want to specify any of the optional parameters.  But in practice - idk, it seems like it'd be a pain that we could save our users.

What do you think of surrounding these lines with a `if (requestBody != null)` check?

Also, this would be a great this to have a test for later on when you get to that point.

File: solr/core/src/java/org/apache/solr/handler/admin/api/ReplaceNodeAPI.java
Comment (gerlowskija) [inline solr/core/src/java/org/apache/solr/handler/admin/api/ReplaceNodeAPI.java:118]: [0] The `@Schema` annotations you see on this code end up used as descriptions in the OpenAPI spec that Solr generates as part of its build.

If you're not familiar with OpenAPI, it's really awesome! It provides tooling you can use to generate a "spec" - a machine-readable description of your API.  And then other tooling can take that spec and do all sorts of things with it, like generating clients for various languages.

Right now, Solr's set up to generate a spec (run `./gradlew resolve` and take a look at `solr/core/build/generated/openapi/openapi.json`).  We haven't started using the spec yet, but it's on my roadmap to set us up to generate Solr clients for Python, golang, etc.

Anyway, this is all to say: these `@Schema` annotations don't seem like they do much now, but they'll be super helpful to have in place once we start using the spec.

You don't need to have them everywhere if you're not sure what to say about certain parameters.  But if you want to include them, you can always copy the snippet in the ref guide that describes the parameter.

**Review by gerlowskija — CHANGES_REQUESTED** [review]
Alright, left a few comments inline, but mostly small changes I think.  Looking good!

Before this gets merged, we'll probably also need some tests, and to update any ref-guide docs that mention the REPLACENODE API, so they have the correct v2 syntax.  Let me know if you have questions about either of those and I can help out as needed.

Thanks Josh!

File: solr/core/src/java/org/apache/solr/handler/admin/CollectionsHandler.java
Comment (joshgog) [reply solr/core/src/java/org/apache/solr/handler/admin/CollectionsHandler.java:1797]: Sure

**Comment by joshgog** [comment]
I have added a test that extends SolrTestCaseJ4 but not sure if its exhaustive

File: solr/core/src/java/org/apache/solr/handler/admin/api/ReplaceNodeAPI.java
Comment (joshgog) [reply solr/core/src/java/org/apache/solr/handler/admin/api/ReplaceNodeAPI.java:119]: Done

**Comment by joshgog** [comment]
> Initially I had tried using JerseyTest but the test failed due to server shutdown.
> 
> ```
> 2> INFO: Started listener bound to [localhost:9998]
>   2> Oct 26, 2022 9:23:11 PM org.glassfish.grizzly.http.server.HttpServer start
>   2> INFO: [HttpServer] Started.
>   2> Oct 26, 2022 9:23:12 PM org.glassfish.grizzly.http.server.NetworkListener shutdownNow
>   2> INFO: Stopped listener bound to [localhost:9998]
> ```
> 
> The code is
> 
> ```
> public class ReplaceNodeAPITest extends JerseyTest {
> 
>   private CoreContainer mockCoreContainer;
>   private static final String sourceNodeName = "demoSourceNode";
>   private static final String targetNodeName = "demoTargetNode";
>   private static final boolean waitForFinalState = false;
>   private SolrQueryRequest mockQueryRequest;
>   private SolrQueryResponse queryResponse;
> 
>   @BeforeClass
>   public static void ensureWorkingMockito() {
>     assumeWorkingMockito();
>   }
> 
>   protected Application configure() {
>     resetMocks();
>     final ResourceConfig config = new ResourceConfig();
>     config.register(ReplaceNodeAPI.class);
>     config.register(SolrJacksonMapper.class);
>     config.register(
>         new AbstractBinder() {
>           @Override
>           protected void configure() {
>             bindFactory(new InjectionFactories.SingletonFactory<>(mockCoreContainer))
>                 .to(CoreContainer.class)
>                 .in(Singleton.class);
>           }
>         });
>     config.register(
>         new AbstractBinder() {
>           @Override
>           protected void configure() {
>             bindFactory(InjectionFactories.SolrQueryRequestFactory.class)
>                 .to(SolrQueryRequest.class)
>                 .in(RequestScoped.class);
>           }
>         });
>     config.register(
>         new AbstractBinder() {
>           @Override
>           protected void configure() {
>             bindFactory(InjectionFactories.SolrQueryResponseFactory.class)
>                 .to(SolrQueryResponse.class)
>                 .in(RequestScoped.class);
>           }
>         });
> 
>     return config;
>   }
> 
>   @Test
>   public void testSuccessfulReplaceNodeCommand() throws Exception {
>     final String expectedJson = "{\"responseHeader\":{\"status\":0,\"QTime\":0}}";
>     final CollectionsHandler collectionsHandler = mock(CollectionsHandler.class);
>     when(mockCoreContainer.getCollectionsHandler()).thenReturn(collectionsHandler);
>     ReplaceNodeAPI.ReplaceNodeRequestBody requestBody = new ReplaceNodeAPI.ReplaceNodeRequestBody(targetNodeName, waitForFinalState);
>     Entity<ReplaceNodeRequestBody> entity = Entity.entity(requestBody, MediaType.APPLICATION_JSON);
>     final Response response =
>         target("cluster/nodes/" + sourceNodeName + "/commands/replace").request().post(entity);
>     final String jsonBody = response.readEntity(String.class);
>     assertEquals(200, response.getStatus());
>     assertEquals("application/json", response.getHeaders().getFirst("Content-type"));
>     assertEquals(expectedJson, "{\"responseHeader\":{\"status\":0,\"QTime\":0}}");
>   }
> 
>   private void resetMocks() {
>     mockCoreContainer = mock(CoreContainer.class);
> 
>   }
> }
> ```

The error received is 500. I think problem is caused by the test case params. I'm working on it

File: solr/core/src/test/org/apache/solr/handler/admin/api/ReplaceNodeAPITest.java
Comment (gerlowskija) [inline solr/core/src/test/org/apache/solr/handler/admin/api/ReplaceNodeAPITest.java:73]: [-1] I think we should either re-work this test a bit to assert on something more meaningful (e.g. the remote-message that gets created from all the API parameters), or we should lean exclusively on the JerseyTest you mentioned in some comments.

Either approach (or some third option if you've got other ideas) is totally fine with me.  But as-is at least this test just doesn't validate all that much.

If you want to go with the first option (reworking this to assert on the remote-message), then AddReplicaPropertyAPITest might be a good reference.  That test makes an API call and uses a special type of mock called an "ArgumentCaptor" to record the "remote message", so that the test can assert on it later.

The key bits in AddReplicaPropertyAPITest are:

```
    messageCapturer = ArgumentCaptor.forClass(ZkNodeProps.class);
```
(creates the argument-captor)

```
    addReplicaPropApi.addReplicaProperty(
        "someColl", "someShard", "someReplica", "somePropName", ANY_REQ_BODY);
    verify(mockCommandRunner).runCollectionCommand(messageCapturer.capture(), any(), anyLong());
```
(calls the new API and "captures" or records the remote-message)

File: solr/core/src/test/org/apache/solr/handler/admin/api/ReplaceNodeAPITest.java
Comment (gerlowskija) [inline solr/core/src/test/org/apache/solr/handler/admin/api/ReplaceNodeAPITest.java:77]: [0] I suspect this line and the one above it are leftover from earlier testing?  At the least they're not doing anything right now - I double-checked ReplaceNodeAPI and it never calls `getCollectionsHandler` or anything similar.

Similarly, the `expectedJson` variable isn't used afaict.

**Comment by joshgog** [comment]
Quite helpful information. Will decide between the Jersey test and the remote message test

File: solr/core/src/java/org/apache/solr/handler/admin/api/ReplaceNodeAPI.java
Comment (gerlowskija) [inline solr/core/src/java/org/apache/solr/handler/admin/api/ReplaceNodeAPI.java:102]: [0] Some of these values we're setting on the remote message might actually be null/unset.  The existing v1 code to create this 'message' goes out of its way to exclude null/unset values.

I'm going to push a small update here to make sure these values are only added to the message  if they're non-null.

**Comment by gerlowskija** [comment]
Hi @joshgog - just wanted to summarize where we are on this.  I think at this point you've addressed everything I (and Noble) brought up in review so far.  Unless I'm missing something?

I've pushed one tiny edit a few minutes ago to ensure that ReplaceNodeAPI doesn't add values to its 'remoteMessage' map if they're null, and some tests for that.  But otherwise this looks good to me.

I'll aim to run tests/'check' on this over the weekend and commit as long as you've done everything here you were looking to?

**Comment by HoustonPutman** [comment]
Not sure if this was discussed, but can we take `/commands` out of the path for this? `/replace` alone is pretty self-explanatory.

**Comment by gerlowskija** [comment]
Gonna hold off on merging this for now based on Houston's comment above...

> can we take /commands out of the path for this? /replace alone is pretty self-explanatory.

Happy to remove it if there's consensus, but `/commands` in the path serves a few functions, at least one of which is probably non-obvious:

1. `/commands` is there for better consistency.  All over the place in the APIs we have this pattern of `/pluralCollectionName/singleInstance`, whether this be in the form of `/collections/techproducts`, or `/shards/shard1`, or `/fields/fieldName`.  "Commands" helps us stay consistent with that pattern - for whatever that's worth.
2. Less-obvious, using `/commands` helps cut down on potential path ambiguity between command names and names of cores, collections, shards, etc.  It's less of an issue with this API in particular, but consider how this plays out for commands on the `/cores` path.  Without "commands" in the path, each command name is a potential collision with a core name specified by a user.  That is: `/api/cores/swap` becomes ambiguous: is a user attempting to swap two cores, or fetch the status of a core called 'swap'.

Anyway, if neither of that sways you @HoustonPutman or @dsmiley , then we can update the API here and others in the spreadsheet more broadly.  But figured i'd make a last defense of it.

**Comment by gerlowskija** [comment]
> the `nodes/SingleNode` is already there, so we are still following this consistency.

Doing it for nodes and not commands or other things is the definition of inconsistency 😛 

We follow it here for the node-name, yes.  But the idea isn't to do this once-per-path and call it a day, it's to do it everywhere we expose a noun or resource in our API paths.  The consistency of design across APIs is what helps users reason about what an API will likely be for some operation they're unfamiliar with.  If a user is familiar with other APIs and has seen Solr use this pattern for nodes, collections, cores, shards, replicas, replica properties, cluster properties, fields, fieldTypes, requestHandlers, and on and on .... well then they'll probably expect something similar here.


IMO, The question shouldn't be whether we can get by without this in the path.  The question should be: what does omitting `/commands` get us that makes taking the hit on consistency worth it?

**Comment by HoustonPutman** [comment]
So Jason and I talked offline and cleared up our misunderstandings 🙂 I didn't know that the API spreadsheet had multiple tabs 😅 

Overall I prefer the consistency of the user experience without `commands/`, though I do now understand the distinction between paths with `commands/` and those without.

I'm just one voice in the sea of contributors, so take my opinion as just that. But I would be willing to go through and help make the changes I'm suggesting (removing `commands/` throughout the API). I wouldn't put that work on anyone else!

**Comment by gerlowskija** [comment]
Ah, I was just coming back here to close the loop but I see Houston beat me to it.

There is a distinction between the APIs that include `/commands` in the path and those that don't, but Houston pointed out (correctly I think) that the distinction might be a bit subtle to expect our users to pick up on.  I suspect David feels similarly.

I think we should go with the majority opinion and drop `/commands` from the path here and elsewhere.

>  I would be willing to go through and help make the changes I'm suggesting (removing commands/ throughout the API). I wouldn't put that work on anyone else!

I appreciate it, but no need.  The change needed here is small and I can just make it now while I've got it pulled up.  That just leaves the spreadsheet, which should be easy enough to CTRL-F on.  I'd appreciate your review on the other tabs in that spreadsheet for sure though, if you've got a few @HoustonPutman.

With that resolved this will be back on my list of things to test and commit.

**Comment by joshgog** [comment]
> I'll aim to run tests/'check' on this over the weekend and commit as long as you've done everything here you were looking to?

I'm trying to catch up after a long weekend. I think all is good after the agreement of removing `/command`

**Review by gerlowskija — APPROVED** [review]

**Comment by joshgog** [comment]
Thank you for the merge

**Pull request merged**
